### PR TITLE
package cloud resource in javaagent - but keep them disabled by default

### PR DESCRIPTION
**Pull request by zeitlinger**
plugins can be enabled or disabled using `otel.resource.providers.<>.enabled`, e.g. `otel.resource.providers.aws.enabled`

Fixes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/10447

### CODE REVIEW

File: .github/renovate.json5 [hunk 1]
```diff
@@ -11,6 +11,17 @@
   // https://github.com/renovatebot/renovate/discussions/8399#discussioncomment-305798
   "separateMinorPatch": true,
   "packageRules": [
+    {
+      "matchPackageNames": [
+        "io.opentelemetry.contrib:opentelemetry-aws-resources",
```
Comment (laurit) [inline .github/renovate.json5:16]: this is already included in some other PR?

File: dependencyManagement/build.gradle.kts [hunk 2]
```diff
@@ -96,7 +96,9 @@ val DEPENDENCIES = listOf(
   "commons-logging:commons-logging:1.3.0",
   "commons-validator:commons-validator:1.8.0",
   "io.netty:netty:3.10.6.Final",
+  "io.opentelemetry.contrib:opentelemetry-aws-resources:1.33.0-alpha",
   "io.opentelemetry.contrib:opentelemetry-aws-xray-propagator:1.28.0-alpha",
+  "io.opentelemetry.contrib:opentelemetry-gcp-resources:1.33.0-alpha",
```
Comment (laurit) [inline dependencyManagement/build.gradle.kts:101]: Could you extract `otelContribVersion` into a variable like we already have for `otelSdkVersion`. This should let renovate update all of these with one PR.

File: javaagent-tooling/build.gradle.kts [hunk 3]
```diff
@@ -37,6 +37,18 @@ dependencies {
 
   implementation("io.opentelemetry.contrib:opentelemetry-aws-xray-propagator")
 
+  implementation("io.opentelemetry.contrib:opentelemetry-aws-resources") {
+    exclude("io.opentelemetry", "opentelemetry-api")
+    exclude("io.opentelemetry", "opentelemetry-sdk")
+    exclude("io.opentelemetry.semconv", "opentelemetry-semconv")
```
Comment (laurit) [inline javaagent-tooling/build.gradle.kts:43]: why is this necessary?

Comment (zeitlinger) [reply javaagent-tooling/build.gradle.kts:43]: it was necessary when I added these to a distro - but I'm not sure what the exact problem was

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Set;
+import javax.annotation.Nullable;
+
+@AutoService(AutoConfigurationCustomizerProvider.class)
+public class ResourceProviderPropertiesCustomizer implements AutoConfigurationCustomizerProvider {
+
+  private static final Set<String> DISABLED_BY_DEFAULT_RESOURCE_PROVIDERS =
+      new HashSet<>(
+          Arrays.asList(
+              "io.opentelemetry.contrib.aws.resource.BeanstalkResourceProvider",
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:27]: I think not all of these need to wait for http requests to timeout or respond. Should we by default disable only the slow ones? I guess for start it doesn't matter.

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  // VisibleForTesting
+  Map<String, String> customize(ConfigProperties config) {
+    Set<String> enabledProviders = new HashSet<>(config.getList(ENABLED_KEY));
+
+    List<String> enabled = new ArrayList<>();
+    List<String> disabled = new ArrayList<>();
+
+    for (String providerName : DISABLED_BY_DEFAULT_RESOURCE_PROVIDERS) {
+      Boolean explictEnabled =
+          config.getBoolean(String.format("otel.instrumentation.%s.enabled", providerName));
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:52]: this is deliberate - as outlined [here](https://github.com/open-telemetry/opentelemetry-java-contrib/issues/1074#issuecomment-1960098684)

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java (hunk 4 above)
Comment (zeitlinger) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:27]: My idea was to do that in a follow up PR - to keep this PR more about the mechanism.

**Review by laurit — COMMENTED** [review]
I also created a PR to add the aws detectors a while back https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7927 That one has a test that verifies that aws provider runs after `ContainerResourceProvider` could you check whether it would make sense to bring that test over to this PR.

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java (hunk 4 above)
Comment (laurit) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:27]: make sense

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (laurit) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:52]: To me this doesn't make sense. User might as well use `otel.java.enabled.resource.providers` to enable these. Using `otel.instrumentation.xxx.enabled` or something similar would make sense to me if `xxx` was something short like `aws` or `gcp` not the full class name and the same flag would enable all the detectors for given platform.

File: javaagent-tooling/build.gradle.kts (hunk 3 above)
Comment (zeitlinger) [reply javaagent-tooling/build.gradle.kts:43]: I double checked - it doesn't make a difference in our case.

**Comment by zeitlinger** [comment]
> I also created a PR to add the aws detectors a while back #7927 That one has a test that verifies that aws provider runs after `ContainerResourceProvider` could you check whether it would make sense to bring that test over to this PR.

@laurit done - and all the other suggestions incorporated

File: .github/renovate.json5 [hunk 6]
```diff
@@ -12,16 +12,11 @@
...
-        "io.opentelemetry.contrib:opentelemetry-aws-xray-propagator",
-        "io.opentelemetry.proto:opentelemetry-proto",
-        "io.opentelemetry.semconv:opentelemetry-semconv"
-      ],
       // Renovate's default behavior is only to update from unstable -> unstable if it's for the
       // major.minor.patch, under the assumption that you would want to update to the stable version
       // of that release instead of the unstable version for a future release
-      // (TODO remove once the artifacts above release stable versions)
-      "ignoreUnstable": false
+      "matchCurrentVersion": "/^.*-alpha$/",
```
Comment (zeitlinger) [inline .github/renovate.json5:18]: see https://github.com/renovatebot/renovate/discussions/13849#discussioncomment-2057657

Comment (zeitlinger) [reply .github/renovate.json5:18]: extracted to standalone PR: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10761

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    List<String> enabled = new ArrayList<>();
+    List<String> disabled = new ArrayList<>();
+
+    for (Map.Entry<String, String> providerEntry :
+        DISABLED_BY_DEFAULT_RESOURCE_PROVIDERS.entrySet()) {
+      String providerName = providerEntry.getKey();
+      String providerGroup = providerEntry.getValue();
+      Boolean explictEnabled =
+          config.getBoolean(String.format("otel.instrumentation.%s.enabled", providerGroup));
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:65]: I'm not sure about using `otel.instrumentation` here as that flag is used to enable/disable instrumentations, but we are dealing with resource providers here.

File: javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizerTest.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.assertj.core.util.Strings;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.TestFactory;
+
+public class ResourceProviderPropertiesCustomizerTest {
+
+  public static final class Provider
+      implements io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider {
```
Comment (laurit) [inline javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizerTest.java:29]: could you import this class

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java (hunk 7 above)
Comment (zeitlinger) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:65]: I think @trask brought up this idea that resource providers are a form of instrumentation.

Maybe `otel.instrumentation.resource-provider.%s.enabled`?

Comment (laurit) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java]: Names are hard, I think it would be best to raise this in the SIG.

Comment (laurit) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java]: SIG proposed using `otel.resource.providers.aws.enabled`

Comment (zeitlinger) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java]: changed @laurit

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    List<String> enabled = new ArrayList<>();
+    List<String> disabled = new ArrayList<>();
+
+    for (Map.Entry<String, String> providerEntry :
+        DISABLED_BY_DEFAULT_RESOURCE_PROVIDERS.entrySet()) {
+      String providerName = providerEntry.getKey();
+      String providerGroup = providerEntry.getValue();
+      Boolean explictEnabled =
+          config.getBoolean(
+              String.format("otel.resource.providers.%s.enabled", providerGroup));
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java:66]: Now the only thing we need to figure out is where to document this. I think it would be ok to create an issue for it if this should be documented in the website repo.

Comment (zeitlinger) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/config/ResourceProviderPropertiesCustomizer.java]: added to https://github.com/open-telemetry/opentelemetry.io/issues/4027

File: .github/renovate.json5 (hunk 6 above)
Comment (zeitlinger) [reply .github/renovate.json5]: done

**Review by laurit — APPROVED** [review]

**Review by trask — COMMENTED** [review]
can you add documentation somewhere (either in this repo, or draft PR to website repo)?
